docs: explanation section for overlays (#370)
canonical/craft-parts: docs/explanation/overlays.rst

****************
``OVERLAY`` Step
****************

> It is sometimes necessary to modify the base filesystem when generating a

Reviewer: I guess what I'm missing here is some kind of definition of "base filesystem" here. Maybe a distinction between this "base" filesystem and the build filesystem where the lifecycle runs?

Author: I've incorporated your suggestion from Sergio's comment above, which should fix both :-)

in the ``${CRAFT_OVERLAY}`` environment variable. For example:

.. code-block:: yaml

    parts:
      no-overlay-access:
        plugin: nil
        override-build: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/empty"
      direct-overlay-access:
        plugin: nil
        overlay-script: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/direct"
      indirect-overlay-access:
        plugin: nil
        after: [direct-overlay-access]
        override-build: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/indirect"

Because ``no-overlay-access`` has no access to the overlay directory, the
``no-overlay-access`` step will fail, as the ``CRAFT_OVERLAY`` environment
variable is unset. Removing that attempted access will make this project
build.

.. _overlay_layers:

Layers
------

Each part has an overlay layer, which acts on a shared storage area in the
processing order of parts. If the part doesn't specify any overlay parameters,
this overlay is empty. The overlay's integrity is checked with a checksum
defined by the following diagram:

.. image:: /images/overlay_checksum.svg
   :alt: Diagram for generating the overlay checksum.

> Each layer combines its parameters with the checksum for the layer below it

Reviewer: This sentence is a little awkward with the many "checksums". Here's a suggestion: `Each layer's checksum is the combination of the layer's properties and the checksum of the layer below it.`

Author: That's a significant improvement. I slightly adjusted your suggestion and then improved the rest of the paragraph too.

and checksums that in order to form its own checksum. The base layer is
checksummed to feed the layer for the first part. As a result, a change to
any layer will require the recalculation of the overlay for all layers
above. The order of layers is determined by the :ref:`part_processing_order`.

Filesystem Mutations
====================

> Mutations to the filesystem may include changes caused by Installation of OS packages (including package database updates), the execution of user scripts in the overlay filesystem context, or file filters.

Reviewer: Long line here

> Mutations to the filesystem may include changes caused by Installation of OS packages (including package database updates), the execution of user scripts in the overlay filesystem context, or file filters.

Reviewer: ```suggestion
Mutations to the filesystem may include changes caused by the installation of OS packages (including package database updates), the execution of user scripts in the overlay filesystem context, or file filters.
```

The outcome of the overlay step for each part includes solely the modifications

> made by that part to underlying layers.. As a consequence, if no modifications

Reviewer: ```suggestion
made by that part to underlying layers. As a consequence, if no modifications
```

are made, the result of the overlay step is empty and the result is the same as
the four-step lifecycle without overlays. Subtractive changes such as file
removals are allowed and handled through special whiteout files conforming to

> the `OCI image layer specification <oci_image_layers>`_.

Reviewer: :+1:

Author: *Content blatantly ripped from Claudio's specification document*

Overlay Processing
------------------

Step Execution
==============

Each layer in the overlay step is generated under the following rules:

#. The overlay step for previous parts must have been executed before processing
   the overlay step for a part.
#. If the part declares no overlay parameters, its layer in the overlay step is
   empty.
#. Otherwise, enable the mechanism that handles filesystem layering, and
   assemble the layer stack up to the part being processed.
#. Install overlay packages on top of the layer stack, and execute the user
   script if defined.
#. Disable the mechanism that handles filesystem layering.
#. Generate the overlay step state.

Staging Overlay Files
=====================

When executing the stage step for a part that declares overlay content, the
consolidated content generated in the overlay step is added to the common stage

> area along with artifacts resulting from the part build step. Files from overlay

Reviewer: ```suggestion
area along with artifacts resulting from the part's build step. Files from overlay
```

****************
``OVERLAY`` Step
****************

> It is sometimes necessary to modify the base filesystem used in when generating

Reviewer 2: > the base filesystem used in when generating

this construct sounds strange to me.

I would also add a ref link to base (if it exists).

Author: Yeah that's because it's just all kinds of awkward :-) I've changed it to "modify the base filesystem when generating a parts-based payload," though I'm open to other suggestions for better ways to introduce overlays.

Unfortunately there's no base ref in the craft-parts doc yet and I couldn't see anything in other projects that would be relevant.

Reviewer: maybe we should explicitly call-out rockcraft here? Something like "Some Craft applications, like Rockcraft, include entire base filesystems in addition to the usual part-generated payload. For those applications, it is often desirable to modify this base filesystem..."
that's not _quite_ right, but something along those paths?

Author: I like that — it anchors the reason for the overlay step to a specific real-world use case so the reader can more quickly decide whether the overlay step is relevant to them.

a parts-based payload. On these occasions, the ``OVERLAY`` step provides the
means to do so.

Overlay Parameters
------------------

A part has three parameters that can be used to adjust how the overlay step
works: ``overlay-packages``, ``overlay-script`` and ``overlay-filter``.
``overlay-packages`` and ``overlay`` (the filter parameter) behave much the
same way as the related parameters on the ``STAGE`` step. ``overlay-script``
likewise behaves similarly to ``override-stage``, including having access to

> the ``partsctl`` command.

Reviewer: do you mean `craftctl`?

An example parts section with overlay parameters looks as follows:

.. code-block:: yaml

    parts:
      part_with_overlay:
        plugin: nil
        overlay_packages:
          - ed
        overlay_script: |
          rm -f ${CRAFT_OVERLAY}/usr/bin/vi ${CRAFT_OVERLAY}/usr/bin/vim*
          rm -f ${CRAFT_OVERLAY}/usr/bin/emacs*
          rm -f ${CRAFT_OVERLAY}/bin/nano
        overlay:
          - bin
          - usr/bin

After running this part, the overlay layer (and the final package) will only
contain ``ed`` as an editor, with vi/vim, emacs, and nano all having been

> removed.

Reviewer 2: ed is monospaced while the other editors are not, what is the criteria?

Author: The criteria comes from me writing the two halves of the sentence 10 minutes apart. Fixed!

.. _overlay_visibility:

Overlay Visibility
------------------

By default, a part does not get access to the filesystem overlay. However,
if a part provides any overlay parameters or depends on another part that
provides overlay parameters, the location of the overlay is made available
in the ``${CRAFT_OVERLAY}`` environment variable. For example:

.. code-block:: yaml

    parts:
      no-overlay-access:
        plugin: nil
        override-build: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/empty"
      direct-overlay-access:
        plugin: nil
        overlay-script: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/direct"
      indirect-overlay-access:
        plugin: nil
        after: [direct-overlay-access]
        override-build: |
          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/indirect"

Because ``no-overlay-access`` has no access to the overlay directory, the

> ``no-overlay-access`` step will fail, as the ``CRAFT_OVERLAY`` environment

Reviewer: ```suggestion
``override-build`` step will fail, as the ``CRAFT_OVERLAY`` environment
```
or possible `build`, as you refer to it as a "step"

Author: Good catch

variable is unset. Removing that attempted access will make this project
build.

Parts are ordered deterministically for a project, guaranteeing that order will
be preserved between runs. The current ordering rules are as follows:

#. Parts are ordered alphabetically by name
#. Any part that is required before another part is moved ahead of it

This means that renaming parts and adding or removing ``after`` sections

> for parts can modify the overlay order.

Reviewer 2: should this not be under an Ordering title? Visibility seems like a stretch

Author: I've moved this over to an expanded part processing order section in the lifecycle explanation, adding a couple of examples. It's now xref'd in the Layers section below.
